### PR TITLE
Added user attributes, client auth resources, service account roles

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**

### CODE REVIEW

**Comment by AndrewChubatiuk** [comment]
@mrparkers 
Hello again. Prepared a PR with some changes
- Added user attributes
- Added client and client authorization policy datasources
- Added client authorization resources (resource, permission, scope)
- Added service account roles resource
Added examples of how to use it

File: keycloak/openid_client.go [hunk 1]
```diff
@@ -1,37 +1,329 @@
...
+
+	for _, resource := range resources {
+		permission.Resources = append(permission.Resources, resource.Id)
+	}
+
+	return &permission, nil
+}
+
+func (keycloakClient *KeycloakClient) NewOpenidClientAuthorizationPermission(permission *OpenidClientAuthorizationPermission) error {
+	body, _, err := keycloakClient.post(fmt.Sprintf("/realms/%s/clients/%s/authz/resource-server/permission", permission.RealmId, permission.ResourceServerId), permission)
```
Comment (mrparkers) [inline keycloak/openid_client.go]: it looks like this was the only reason the `keycloakClient.post` signature changed.  Could you help me understand why the response body is important here?

File: provider/keycloak_openid_client_optional_scopes_test.go [hunk 2]
```diff
@@ -426,6 +424,7 @@ resource "keycloak_openid_client_optional_scopes" "optional_scopes" {
 		"address",
 		"phone",
 		"offline_access",
+		"microprofile-jwt",
```
Comment (mrparkers) [inline provider/keycloak_openid_client_optional_scopes_test.go]: I assume this optional scope is now added to every client by default as of 5.0.0 or 6.0.0?  I'm still on 4.8.3.Final.

File: provider/keycloak_openid_client_service_account_role.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	"github.com/mrparkers/terraform-provider-keycloak/keycloak"
+	"strings"
+)
+
+func resourceKeycloakOpenidClientServiceAccountRole() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceKeycloakOpenidClientServiceAccountRoleCreate,
+		Read:   resourceKeycloakOpenidClientServiceAccountRoleRead,
+		Delete: resourceKeycloakOpenidClientServiceAccountRoleDelete,
+		// This resource can be imported using {{realm}}/{{client_id}}. The Client ID is displayed in the GUI
```
Comment (mrparkers) [inline provider/keycloak_openid_client_service_account_role.go]: This comment doesn't seem correct.  You can just remove it if you want.

File: provider/keycloak_user.go [hunk 4]
```diff
@@ -49,6 +49,30 @@ func resourceKeycloakUser() *schema.Resource {
 				Type:     schema.TypeString,
 				Optional: true,
 			},
+			"attributes": {
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+			"federated_identity": {
```
Comment (mrparkers) [inline provider/keycloak_user.go]: Should this be `Computed`?  My understanding is that this only exists if the user was federated from an identity provider.

File: provider/keycloak_openid_client_authorization_scope.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+	"strings"
+)
+
+func resourceKeycloakOpenidClientAuthorizationScope() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceKeycloakOpenidClientAuthorizationScopeCreate,
+		Read:   resourceKeycloakOpenidClientAuthorizationScopeRead,
+		Delete: resourceKeycloakOpenidClientAuthorizationScopeDelete,
+		Update: resourceKeycloakOpenidClientAuthorizationScopeUpdate,
+		// This resource can be imported using {{realm}}/{{resource_server_id}}. The Client ID is displayed in the GUI
```
Comment (mrparkers) [inline provider/keycloak_openid_client_authorization_scope.go]: This comment doesn't seem correct.  You can just remove it if you want.

File: provider/keycloak_openid_client_authorization_resource.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+	"strings"
+)
+
+func resourceKeycloakOpenidClientAuthorizationResource() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceKeycloakOpenidClientAuthorizationResourceCreate,
+		Read:   resourceKeycloakOpenidClientAuthorizationResourceRead,
+		Delete: resourceKeycloakOpenidClientAuthorizationResourceDelete,
+		Update: resourceKeycloakOpenidClientAuthorizationResourceUpdate,
+		// This resource can be imported using {{realm}}/{{resource_server_id}}. The Client ID is displayed in the GUI
```
Comment (mrparkers) [inline provider/keycloak_openid_client_authorization_resource.go]: This comment doesn't seem correct.  You can just remove it if you want.

File: provider/keycloak_openid_client_authorization_resource.go [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"display_name": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"uris": {
+				Type:     schema.TypeList,
```
Comment (mrparkers) [inline provider/keycloak_openid_client_authorization_resource.go]: Does the order of `uris` matter?  If not, this should be a `TypeSet`.

File: provider/keycloak_openid_client_authorization_permission.go [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+	keycloakOpenidClientPermissionTypes                      = []string{"resource", "scope"}
+)
+
+func resourceKeycloakOpenidClientAuthorizationPermission() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceKeycloakOpenidClientAuthorizationPermissionCreate,
+		Read:   resourceKeycloakOpenidClientAuthorizationPermissionRead,
+		Delete: resourceKeycloakOpenidClientAuthorizationPermissionDelete,
+		Update: resourceKeycloakOpenidClientAuthorizationPermissionUpdate,
+		// This resource can be imported using {{realm}}/{{resource_server_id}}. The Client ID is displayed in the GUI
```
Comment (mrparkers) [inline provider/keycloak_openid_client_authorization_permission.go]: This comment doesn't seem correct.  You can just remove it if you want.

File: provider/keycloak_openid_client_authorization_permission.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+				ValidateFunc: validation.StringInSlice(keycloakOpenidClientResourcePermissionDecisionStrategies, false),
+				Default:      "UNANIMOUS",
+			},
+			"policies": {
+				Type:     schema.TypeList,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+				Optional: true,
+			},
+			"resources": {
+				Type:     schema.TypeList,
```
Comment (mrparkers) [inline provider/keycloak_openid_client_authorization_permission.go]: Does the order of `resources` matter?  If not, this should be a `TypeSet`.

File: provider/keycloak_openid_client_authorization_permission.go [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+				Optional: true,
+			},
+			"decision_strategy": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringInSlice(keycloakOpenidClientResourcePermissionDecisionStrategies, false),
+				Default:      "UNANIMOUS",
+			},
+			"policies": {
+				Type:     schema.TypeList,
```
Comment (mrparkers) [inline provider/keycloak_openid_client_authorization_permission.go]: Does the order of `policies` matter?  If not, this should be a `TypeSet`.

File: provider/keycloak_openid_client.go [hunk 11]
```diff
@@ -88,11 +85,43 @@ func resourceKeycloakOpenidClient() *schema.Resource {
 				Set:      schema.HashString,
 				Optional: true,
 			},
+			"service_accounts_enabled": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+			},
+			"service_account_user_id": {
```
Comment (mrparkers) [inline provider/keycloak_openid_client.go]: Nice change 👍 - I've been meaning to add this.

File: provider/keycloak_openid_client_test.go [hunk 12]
```diff
@@ -129,6 +129,10 @@ func TestAccKeycloakOpenidClient_updateInPlace(t *testing.T) {
 	directAccessGrantsEnabled := randomBool()
 	serviceAccountsEnabled := randomBool()
 
+	if !standardFlowEnabled {
+		implicitFlowEnabled = !standardFlowEnabled
+	}
```
Comment (mrparkers) [inline provider/keycloak_openid_client_test.go]: What's the reason for this change?

File: provider/keycloak_openid_client.go [hunk 13]
```diff
@@ -211,7 +298,17 @@ func resourceKeycloakOpenidClientUpdate(data *schema.ResourceData, meta interfac
...
+
+	if client.ServiceAccountsEnabled {
+		serviceAccountUser, err := keycloakClient.GetOpenidClientServiceAccountUserId(client.RealmId, client.Id)
+		if err != nil {
+			return err
+		}
+		serviceAccountUserId = serviceAccountUser.Id
+	}
+
+	setOpenidClientData(data, client, serviceAccountUserId)
```
Comment (mrparkers) [inline provider/keycloak_openid_client.go]: This block of code (lines 301 - 311) seems to be repeated multiple times.  Could you update the `setOpenidClientData` to do this instead?

File: keycloak/openid_client.go [hunk 14]
```diff
@@ -1,37 +1,329 @@
...
+
+func (keycloakClient *KeycloakClient) GetOpenidClientServiceAccountRole(realm, serviceAccountUserId, clientId, roleId string) (*OpenidClientServiceAccountRole, error) {
+	serviceAccountRoles := []OpenidClientServiceAccountRole{}
+	serviceAccountRoles = append(serviceAccountRoles, OpenidClientServiceAccountRole{
+		Id:                   roleId,
+		RealmId:              realm,
+		ContainerId:          clientId,
+		ServiceAccountUserId: serviceAccountUserId,
+	})
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/users/%s/role-mappings/clients/%s", realm, serviceAccountUserId, clientId), &serviceAccountRoles, nil)
```
Comment (mrparkers) [inline keycloak/openid_client.go]: I'm not sure I understand what's going on here.  It looks like you're initializing the `serviceAccountRoles` slice with a single element, but then it's immediately being overwritten by this line.  Is there a reason for this?

File: keycloak/openid_client.go [hunk 15]
```diff
@@ -1,37 +1,329 @@
...
+	_, _, err = keycloakClient.post(fmt.Sprintf("/realms/%s/users/%s/role-mappings/clients/%s", serviceAccountRole.RealmId, serviceAccountRole.ServiceAccountUserId, serviceAccountRole.ContainerId), serviceAccountRoles)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (keycloakClient *KeycloakClient) DeleteOpenidClientServiceAccountRole(realm, serviceAccountUserId, clientId, roleId string) error {
+	serviceAccountRoles := []OpenidClientServiceAccountRole{}
+	serviceAccountRoles = append(serviceAccountRoles, OpenidClientServiceAccountRole{
```
Comment (mrparkers) [inline keycloak/openid_client.go]: Same comment as above.

File: keycloak/openid_client.go [hunk 16]
```diff
@@ -1,37 +1,329 @@
...
+}
+
+func (keycloakClient *KeycloakClient) NewOpenidClientServiceAccountRole(serviceAccountRole *OpenidClientServiceAccountRole) error {
+	role, err := keycloakClient.GetClientRoleByName(serviceAccountRole.RealmId, serviceAccountRole.ContainerId, serviceAccountRole.Name)
+	if err != nil {
+		return err
+	}
+	serviceAccountRole.Id = role.Id
+	serviceAccountRoles := []OpenidClientServiceAccountRole{}
+	serviceAccountRoles = append(serviceAccountRoles, *serviceAccountRole)
```
Comment (mrparkers) [inline keycloak/openid_client.go]: Do you need to use `append` here?  I think you could just initialize the slice with the single element instead of making it empty and adding to it.

File: keycloak/openid_client.go [hunk 17]
```diff
@@ -1,37 +1,329 @@
...
+	Name                 string `json:"name"`
+	ClientRole           bool   `json:"clientRole"`
+	Composite            bool   `json:"composite"`
+	ContainerId          string `json:"containerId"`
+}
+
+type OpenidClientAuthorizationSettings struct {
+	PolicyEnforcementMode         string `json:"policyEnforcementMode,omitempty"`
+	AllowRemoteResourceManagement bool   `json:"allowRemoteResourceManagement,omitempty"`
+}
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply keycloak/openid_client.go]: It's needed to get permission Id from response body. ID cannot be extracted from location header as it's done for the rest of the resources

File: keycloak/openid_client.go (hunk 1 above)
Comment (AndrewChubatiuk) [reply keycloak/openid_client.go]: It's needed to get permission id from response body. Id cannot be extracted from location header as it's done for the rest of resources

File: keycloak/group.go [hunk 18]
```diff
@@ -114,13 +114,13 @@ func (keycloakClient *KeycloakClient) UpdateGroup(group *Group) error {
...
 func (keycloakClient *KeycloakClient) DeleteGroup(realmId, id string) error {
-	return keycloakClient.delete(fmt.Sprintf("/realms/%s/groups/%s", realmId, id))
+	return keycloakClient.delete(fmt.Sprintf("/realms/%s/groups/%s", realmId, id), nil)
 }
 
 func (keycloakClient *KeycloakClient) ListGroupsWithName(realmId, name string) ([]*Group, error) {
 	var groups []*Group
 
-	err := keycloakClient.get(fmt.Sprintf("/realms/%s/groups?search=%s", realmId, url.QueryEscape(name)), &groups)
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/groups?search=%s", realmId, url.QueryEscape(name)), &groups, nil)
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply keycloak/group.go]: Done

File: keycloak/openid_client.go [hunk 19]
```diff
@@ -86,9 +378,9 @@ func (keycloakClient *KeycloakClient) GetOpenidClient(realmId, id string) (*Open
 
 func (keycloakClient *KeycloakClient) GetOpenidClientByClientId(realmId, clientId string) (*OpenidClient, error) {
 	var clients []OpenidClient
-	var clientSecret openidClientSecret
+	var clientSecret OpenidClientSecret
 
-	err := keycloakClient.get(fmt.Sprintf("/realms/%s/clients?clientId=%s", realmId, clientId), &clients)
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/clients?clientId=%s", realmId, clientId), &clients, nil)
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply keycloak/openid_client.go]: Done

File: keycloak/user.go [hunk 20]
```diff
@@ -65,13 +75,13 @@ func (keycloakClient *KeycloakClient) UpdateUser(user *User) error {
...
 func (keycloakClient *KeycloakClient) DeleteUser(realmId, id string) error {
-	return keycloakClient.delete(fmt.Sprintf("/realms/%s/users/%s", realmId, id))
+	return keycloakClient.delete(fmt.Sprintf("/realms/%s/users/%s", realmId, id), nil)
 }
 
 func (keycloakClient *KeycloakClient) GetUserByUsername(realmId, username string) (*User, error) {
 	var users []*User
 
-	err := keycloakClient.get(fmt.Sprintf("/realms/%s/users?username=%s", realmId, url.QueryEscape(username)), &users)
+	err := keycloakClient.get(fmt.Sprintf("/realms/%s/users?username=%s", realmId, url.QueryEscape(username)), &users, nil)
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply keycloak/user.go]: Done

File: keycloak/openid_client.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply keycloak/openid_client.go]: Done

File: provider/keycloak_openid_client_service_account_role.go (hunk 3 above)
Comment (AndrewChubatiuk) [reply provider/keycloak_openid_client_service_account_role.go]: Removed

File: provider/keycloak_openid_client_authorization_scope.go (hunk 5 above)
Comment (AndrewChubatiuk) [reply provider/keycloak_openid_client_authorization_scope.go]: Done

File: keycloak/openid_client.go (hunk 16 above)
Comment (AndrewChubatiuk) [reply keycloak/openid_client.go]: Done

File: keycloak/openid_client.go (hunk 15 above)
Comment (AndrewChubatiuk) [reply keycloak/openid_client.go]: Done

File: keycloak/openid_client.go (hunk 14 above)
Comment (AndrewChubatiuk) [reply keycloak/openid_client.go]: Fixed

File: provider/keycloak_openid_client_test.go (hunk 12 above)
Comment (AndrewChubatiuk) [reply provider/keycloak_openid_client_test.go]: At least standard or implicit flow should be enabled to set valid redirect URI

File: provider/keycloak_openid_client.go (hunk 13 above)
Comment (AndrewChubatiuk) [reply provider/keycloak_openid_client.go]: Done

File: provider/keycloak_openid_client_authorization_permission.go (hunk 8 above)
Comment (AndrewChubatiuk) [reply provider/keycloak_openid_client_authorization_permission.go]: Removed

File: provider/keycloak_openid_client_authorization_permission.go (hunk 9 above)
Comment (AndrewChubatiuk) [reply provider/keycloak_openid_client_authorization_permission.go]: Done

File: provider/keycloak_openid_client_authorization_permission.go (hunk 10 above)
Comment (AndrewChubatiuk) [reply provider/keycloak_openid_client_authorization_permission.go]: Done

File: provider/data_source_keycloak_openid_client.go [hunk 21]
```diff
@@ -0,0 +1,123 @@
+package provider
+
+import (
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/mrparkers/terraform-provider-keycloak/keycloak"
+)
+
+func dataSourceKeycloakOpenidClient() *schema.Resource {
+	return &schema.Resource{
```
[the remark this answers is not in the document]
Comment (AndrewChubatiuk) [reply provider/data_source_keycloak_openid_client.go]: Right, now the only difference is that data source attributes are computed, but usually data source scheme has bigger differences. I would like to leave it as it is

File: provider/keycloak_openid_client_test.go (hunk 12 above)
Comment (mrparkers) [reply provider/keycloak_openid_client_test.go]: Is that new behavior with version 5.0.0 or 6.0.0 of Keycloak.  I realize that there is no point in setting valid redirect uris without either of these two flows enabled, but I don't believe Keycloak 4.8.3 requires it.

I'm fine with keeping this change, I was just curious what the context was.

**Comment by AndrewChubatiuk** [comment]
@mrparkers 
Fixed all your PR comments and added tests.
